File: frontend/src/types/vulns.ts
/**
 * Supply-chain read API (broker, PR #1671): vulnerabilities, SBOMs and the
 * image inventory they join to. Shapes follow the captured responses in
 * fixtures/vuln-captures.
 *
 * Conventions the UI keeps:
 *  - `null` is unknown, never "no" / "safe" / zero (kev, epss, exposed,
 *    inUse, sbomTrust, dbUpdatedAt ...);
 *  - an empty `reports` list means no vulnerability data for the image:
 *    unknown, not clean;
 *  - `title` / `primaryUrl` are third-party text: rendered as text or a
 *    plain link, never fetched or executed.
 */

export type VulnSeverity = 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW' | 'NONE' | 'UNKNOWN';
export type VulnSource = 'trivy-operator' | 'grype' | 'registry';
/** How a report was matched to what runs, strongest first. */
export type JoinKind = 'image_id' | 'platform_manifest' | 'workload_tag' | 'report_digest';
/** Weakest first. Only `verified` may be shown as signed. */
export type SbomTrust = 'attached-unbound' | 'unverified' | 'scanned' | 'verified';
/** `executed` | `loaded` | `unknown` | `installed_not_observed` (#1678); `unknown` before it. */
export type InUseState = 'executed' | 'loaded' | 'unknown' | 'installed_not_observed' | string;

/**
 * Fields added by the in-use tiers release (#1678). All optional: an older
 * Broker omits them, and the UI then shows the tier as unknown.
 */
export interface InUseDetail {
  state: InUseState;
  /** Why the state is unknown (language_package, no_runtime_data, capture_gap, host_network, no_package_files). */
  reason: string | null;
  observedSince: string | null;
  windowHours: number;
  containers: number;
  /** `file` | `static_binary` | `interpreted`. */
  coverage: string;
}

// ── GET /vulnerabilities ────────────────────────────────────────────────
export interface CveSummary {
  id: string;
  severity: VulnSeverity;
  maxScore: number | null;
  fixable: boolean;
  kev: boolean | null;
  maxEpss: number | null;
  packages: string[];
  sources: string[];
  images: number;
  workloads: number;
  runningWorkloads: number;
  namespaces: number;
  weakestJoin: JoinKind;
  inUse: boolean | null;
  inUseState: InUseState;
  /** #1678: the most urgent tier over every affected workload container in scope. */
  tier?: string;
  executedWorkloads?: number;
  loadedWorkloads?: number;
  unknownWorkloads?: number;
  notObservedWorkloads?: number;
  exposedWorkloads?: number;
}

export interface CvePage {
  items: CveSummary[];
  nextAfter: string | null;
  /** When the summary was last rebuilt; null until the first rebuild. */
  computedAt: string | null;
  staleSeconds: number | null;
}

// ── GET /vulnerabilities/{id}/exposure ───────────────────────────────────
export interface ExposedPackage {
  name: string;
  installedVersion: string;
  fixedVersions: string[];
  severity: VulnSeverity;
  sources: string[];
}

export interface ExposedImage {
  digest: string;
  repository: string | null;
  tags: string[];
  sources: string[];
  reportDigests: string[];
  join: JoinKind;
  severity: VulnSeverity;
  packages: ExposedPackage[];
}

export type ExposedVia = 'other_namespace' | 'unattributed' | 'public_ip' | 'node';

export interface NetworkExposure {
  windowHours: number;
  podsObserved: number;
  flowsObserved: number;
  ingressFlowsObserved: number;
  ingressFromOtherNamespaces: number;
  ingressFromUnattributedPeers: number;
  ingressFromPublicIps: number;
  ingressFromNodes: number;
  /** true: outside ingress seen; false: ingress seen, none from outside; null: no ingress observed (unknown). */
  exposed: boolean | null;
  exposedVia: ExposedVia[];
}

export interface ExposedWorkload {
  clusterId: string;
  namespace: string;
  kind: string;
  name: string;
  container: string;
  imageDigest: string;
  join: JoinKind;
  running: boolean;
  lastSeen: string;
  network: NetworkExposure | null;
  inUse: boolean | null;
  inUseState: InUseState;
}

export interface NamespaceExposure {
  namespace: string;
  workloads: number;
  runningWorkloads: number;
  exposedWorkloads: number;
  unknownExposureWorkloads: number;
}

export interface Exposure {
  id: string;
  severity: VulnSeverity;
  fixable: boolean;
  truncated: boolean;
  images: ExposedImage[];
  workloads: ExposedWorkload[];
  namespaces: NamespaceExposure[];
  inUse: boolean | null;
  inUseState: InUseState;
}

// ── GET /images/{digest}/vulnerabilities and /sbom ───────────────────────
export interface Report {
  source: VulnSource | string;
  reportDigest: string;
  join: JoinKind;
  digestKind: string;
  scannedAt: string;
  dbUpdatedAt: string | null;
  scannerName: string | null;
  scannerVersion: string | null;
  osFamily: string | null;
  osName: string | null;
  osEosl: boolean;
  imageRef: string | null;
  itemCount: number;
  sbomFormat: string | null;
  receivedAt: string;
  sbomSources: string[];
  /** null = not stated by the source. */
  sbomTrust: SbomTrust | null;
  attestation: {
    mechanism: string | null;
    artifact_digest: string | null;
    media_type: string | null;
    predicate_type: string | null;
    verified: boolean | null;
  } | null;
}

export interface Finding {
  id: string;
  package: { name: string; type: string | null; purl: string | null };
  installedVersion: string;
  /** Every distinct fixed version the sources give, ordered by source. */
  fixedVersions: string[];
  fixable: boolean;
  severity: VulnSeverity;
  score: number | null;
  cvss: unknown;
  title: string | null;
  primaryUrl: string | null;
  target: string | null;
  class: string | null;
  publishedAt: string | null;
  lastModifiedAt: string | null;
  filePaths: string[];
  kev: boolean | null;
  kevDateAdded: string | null;
  epss: number | null;
  epssPercentile: number | null;
  sources: string[];
  reportDigests: string[];
  inUse: boolean | null;
  inUseState: InUseState;
  /** #1678. Worst over every workload container running the image. */
  tier?: string;
  /** #1678: what produced `tier`, e.g. ["in_use:loaded", "kev", "severity:high", "exposed"]. */
  tierFactors?: string[];
  inUseDetail?: InUseDetail;
}

export interface ImageVulnsPage {
  digest: string;
  /** Empty = no vulnerability data for this image (unknown, not clean). */
  reports: Report[];
  items: Finding[];
  nextAfter: string | null;
}

export interface SbomComponent {
  id: number;
  name: string;
  version: string | null;
  purl: string | null;
  type: string | null;
  class: string | null;
  licenses: string[];
  srcName: string | null;
  srcVersion: string | null;
  layerDigest: string | null;
  filePaths: string[];
}

export interface SbomPage {
  digest: string;
  /** Every source's SBOM with its trust. */
  reports: Report[];
  /** The one the items come from; null = no SBOM. */
  report: Report | null;
  items: SbomComponent[];
  /** Component id cursor (numeric, unlike the other reads' string cursors). */
  nextAfter: number | null;
}

// ── GET /images and /images/{digest} (image inventory, #1655) ────────────
export interface ImageSummary {
  digest: string;
  repository: string | null;
  tags: string[];
  digestKind: string;
  firstSeen: string;
  lastSeen: string;
  runningContainers: number;
}

export interface ImagePage {
  items: ImageSummary[];
  nextAfter: string | null;
}

export interface ImageUser {
  clusterId: string;
  namespace: string;
  workloadKind: string;
  workloadName: string;
  containerName: string;
  containerKind: string;
  imageRef: string;
  firstSeen: string;
  lastSeen: string;
  state: string | null;
  stateReason: string | null;
  ranAsInit: boolean;
  running: boolean;
}

export interface ImageDetail extends Omit<ImageSummary, 'runningContainers'> {
  workloads: ImageUser[];
  truncated: boolean;
}

File: advisor/pkg/api/vulns.go
package api

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strconv"

	log "github.com/rs/zerolog/log"
)

// Vulnerability and SBOM reads (broker supply-chain API): findings the
// supplychain component ingested from Trivy Operator reports and, when the
// opt-in matcher is enabled, kguardian's own Grype matches over SBOMs,
// keyed by image digest.
//
// null means unknown throughout: an image with no report has no data (not
// "no vulnerabilities"), kev/epss null means no source said, and inUse is
// null until kguardian can tell which packages a workload loads.

// VulnReport is one source's report on an image.
type VulnReport struct {
	Source         string   `json:"source"`
	ReportDigest   string   `json:"reportDigest"`
	Join           string   `json:"join"`
	DigestKind     string   `json:"digestKind"`
	ScannedAt      string   `json:"scannedAt"`
	DBUpdatedAt    *string  `json:"dbUpdatedAt"`
	ScannerName    *string  `json:"scannerName"`
	ScannerVersion *string  `json:"scannerVersion"`
	OSFamily       *string  `json:"osFamily"`
	OSName         *string  `json:"osName"`
	ItemCount      int      `json:"itemCount"`
	SbomFormat     *string  `json:"sbomFormat"`
	SbomSources    []string `json:"sbomSources"`
	SbomTrust      *string  `json:"sbomTrust"`
	Attestation    *struct {
		Mechanism     *string `json:"mechanism"`
		PredicateType *string `json:"predicate_type"`
		Verified      bool    `json:"verified"`
	} `json:"attestation"`
}

// VulnPackage names the affected package.
type VulnPackage struct {
	Name string  `json:"name"`
	Type *string `json:"type"`
	Purl *string `json:"purl"`
}

// VulnFinding is one finding, deduplicated across sources.
type VulnFinding struct {
	ID               string      `json:"id"`
	Package          VulnPackage `json:"package"`
	InstalledVersion string      `json:"installedVersion"`
	// FixedVersions lists every distinct fixed version the sources give,
	// in source order (not version order: "10.1" sorts before "9.2").
	FixedVersions []string `json:"fixedVersions"`
	Fixable       bool     `json:"fixable"`
	Severity      string   `json:"severity"`
	Score         *float64 `json:"score"`
	Title         *string  `json:"title"`
	Kev           *bool    `json:"kev"`
	Epss          *float64 `json:"epss"`
	Sources       []string `json:"sources"`
	InUse         *bool    `json:"inUse"`
	InUseState    string   `json:"inUseState"`
}

// ImageVulnsPage is GET /images/{digest}/vulnerabilities.
type ImageVulnsPage struct {
	Digest    string        `json:"digest"`
	Reports   []VulnReport  `json:"reports"`
	Items     []VulnFinding `json:"items"`
	NextAfter *string       `json:"nextAfter"`
}

// ImageVulnsOptions filters GET /images/{digest}/vulnerabilities. Zero
// values are not sent; Fixable nil = both.
type ImageVulnsOptions struct {
	Severity string
	Fixable  *bool
	Source   string
	Limit    int
	After    string
}

// CveSummary is one row of GET /vulnerabilities.
type CveSummary struct {
	ID               string   `json:"id"`
	Severity         string   `json:"severity"`
	MaxScore         *float64 `json:"maxScore"`
	Fixable          bool     `json:"fixable"`
	Kev              *bool    `json:"kev"`
	MaxEpss          *float64 `json:"maxEpss"`
	Packages         []string `json:"packages"`
	Sources          []string `json:"sources"`
	Images           int64    `json:"images"`
	Workloads        int64    `json:"workloads"`
	RunningWorkloads int64    `json:"runningWorkloads"`
	Namespaces       int64    `json:"namespaces"`
	WeakestJoin      string   `json:"weakestJoin"`
	InUse            *bool    `json:"inUse"`
}

// CvePage is GET /vulnerabilities. ComputedAt nil = the summary has not
// been built since the broker started (unknown, not empty).
type CvePage struct {
	Items        []CveSummary `json:"items"`
	NextAfter    *string      `json:"nextAfter"`
	ComputedAt   *string      `json:"computedAt"`
	StaleSeconds *int64       `json:"staleSeconds"`
}

// VulnsListOptions filters GET /vulnerabilities.
type VulnsListOptions struct {
	Namespace string
	Severity  string
	Fixable   *bool
	Running   bool
	Limit     int
	After     string
}

// NetworkExposure is the observed ingress for one workload. Exposed is
// false only when ingress was observed (IngressFlowsObserved > 0) and none
// came from outside; nil = unknown (no ingress observed, even with egress:
// inbound UDP is not captured), never "not exposed".
type NetworkExposure struct {
	WindowHours                  int64    `json:"windowHours"`
	PodsObserved                 int64    `json:"podsObserved"`
	FlowsObserved                int64    `json:"flowsObserved"`
	IngressFlowsObserved         int64    `json:"ingressFlowsObserved"`
	IngressFromOtherNamespaces   int64    `json:"ingressFromOtherNamespaces"`
	IngressFromUnattributedPeers int64    `json:"ingressFromUnattributedPeers"`
	IngressFromPublicIPs         int64    `json:"ingressFromPublicIps"`
	IngressFromNodes             int64    `json:"ingressFromNodes"`
	Exposed                      *bool    `json:"exposed"`
	ExposedVia                   []string `json:"exposedVia"`
}

// ExposedImage is an inventory image a CVE affects.
type ExposedImage struct {
	Digest     string   `json:"digest"`
	Repository *string  `json:"repository"`
	Tags       []string `json:"tags"`
	Sources    []string `json:"sources"`
	Join       string   `json:"join"`
	Severity   string   `json:"severity"`
	Packages   []struct {
		Name             string   `json:"name"`
		InstalledVersion string   `json:"installedVersion"`
		FixedVersions    []string `json:"fixedVersions"`
		Severity         string   `json:"severity"`
	} `json:"packages"`
}

// ExposedWorkload is a workload container running (or having run) an
// affected image.
type ExposedWorkload struct {
	Namespace   string          `json:"namespace"`
	Kind        string          `json:"kind"`
	Name        string          `json:"name"`
	Container   string          `json:"container"`
	ImageDigest string          `json:"imageDigest"`
	Join        string          `json:"join"`
	Running     bool            `json:"running"`
	LastSeen    string          `json:"lastSeen"`
	Network     NetworkExposure `json:"network"`
	InUse       *bool           `json:"inUse"`
}

// NamespaceExposure summarises one namespace.
type NamespaceExposure struct {
	Namespace                string `json:"namespace"`
	Workloads                int64  `json:"workloads"`
	RunningWorkloads         int64  `json:"runningWorkloads"`
	ExposedWorkloads         int64  `json:"exposedWorkloads"`
	UnknownExposureWorkloads int64  `json:"unknownExposureWorkloads"`
}

// Exposure is GET /vulnerabilities/{id}/exposure.
type Exposure struct {
	ID         string              `json:"id"`
	Severity   string              `json:"severity"`
	Fixable    bool                `json:"fixable"`
	Images     []ExposedImage      `json:"images"`
	Workloads  []ExposedWorkload   `json:"workloads"`
	Namespaces []NamespaceExposure `json:"namespaces"`
	Truncated  bool                `json:"truncated"`
	InUse      *bool               `json:"inUse"`
}

// SbomComponent is one SBOM component.
type SbomComponent struct {
	ID       int64    `json:"id"`
	Name     string   `json:"name"`
	Version  *string  `json:"version"`
	Purl     *string  `json:"purl"`
	Type     *string  `json:"type"`
	Licenses []string `json:"licenses"`
}

// SbomPage is GET /images/{digest}/sbom. Report nil = no SBOM (unknown).
type SbomPage struct {
	Digest    string          `json:"digest"`
	Reports   []VulnReport    `json:"reports"`
	Report    *VulnReport     `json:"report"`
	Items     []SbomComponent `json:"items"`
	NextAfter *int64          `json:"nextAfter"`
}

// maxCycloneDXBytes bounds the CycloneDX download. The broker caps the
// document at its SBOM component limit; this is a client-side backstop
// well above that, and a document that reaches it is refused rather than
// written out truncated.
const maxCycloneDXBytes = 64 << 20

// Swappable for tests that bypass HTTP.
var (
	GetImageVulnsFunc    = getRealImageVulns
	GetVulnsFunc         = getRealVulns
	GetExposureFunc      = getRealExposure
	GetSbomFunc          = getRealSbom
	GetSbomCycloneDXFunc = getRealSbomCycloneDX
)

// GetImageVulns returns one page of findings for an image, plus the raw body.
func GetImageVulns(digest string, opts ImageVulnsOptions) (*ImageVulnsPage, []byte, error) {
	return GetImageVulnsFunc(digest, opts)
}

// GetVulns returns one page of the cluster CVE summary, plus the raw body.
func GetVulns(opts VulnsListOptions) (*CvePage, []byte, error) { return GetVulnsFunc(opts) }

// GetExposure returns where a CVE runs and its observed exposure. 0 =
// broker default window.
func GetExposure(id string, windowHours int) (*Exposure, []byte, error) {
	return GetExposureFunc(id, windowHours)
}

// GetSbom returns one page of an image's SBOM components.
func GetSbom(digest, source string, limit int, after int64) (*SbomPage, []byte, error) {
	return GetSbomFunc(digest, source, limit, after)
}

// GetSbomCycloneDX returns the broker's CycloneDX export for an image.
func GetSbomCycloneDX(digest, source string) ([]byte, error) {
	return GetSbomCycloneDXFunc(digest, source)
}

func boolQuery(q url.Values, key string, v *bool) {
	if v != nil {
		q.Set(key, strconv.FormatBool(*v))
	}
}

func withQuery(path string, q url.Values) string {
	if enc := q.Encode(); enc != "" {
		return path + "?" + enc
	}
	return path
}

func decodeInto[T any](op string, body []byte) (*T, []byte, error) {
	var out T
	if err := json.Unmarshal(body, &out); err != nil {
		return nil, nil, fmt.Errorf("%s: decoding response: %w", op, err)
	}
	return &out, body, nil
}

func getRealImageVulns(digest string, o ImageVulnsOptions) (*ImageVulnsPage, []byte, error) {
	q := url.Values{}
	if o.Severity != "" {
		q.Set("severity", o.Severity)
	}
	boolQuery(q, "fixable", o.Fixable)
	if o.Source != "" {
		q.Set("source", o.Source)
	}
	if o.Limit > 0 {
		q.Set("limit", strconv.Itoa(o.Limit))
	}
	if o.After != "" {
		q.Set("after", o.After)
	}
	body, err := brokerGetBody("GetImageVulns", withQuery("/images/"+url.PathEscape(digest)+"/vulnerabilities", q))
	if err != nil {
		return nil, nil, err
	}
	return decodeInto[ImageVulnsPage]("GetImageVulns", body)
}

func getRealVulns(o VulnsListOptions) (*CvePage, []byte, error) {
	q := url.Values{}
	if o.Namespace != "" {
		q.Set("namespace", o.Namespace)
	}
	if o.Severity != "" {
		q.Set("severity", o.Severity)
	}
	boolQuery(q, "fixable", o.Fixable)
	if o.Running {
		q.Set("running", "true")
	}
	if o.Limit > 0 {
		q.Set("limit", strconv.Itoa(o.Limit))
	}
	if o.After != "" {
		q.Set("after", o.After)
	}
	body, err := brokerGetBody("GetVulns", withQuery("/vulnerabilities", q))
	if err != nil {
		return nil, nil, err
	}
	return decodeInto[CvePage]("GetVulns", body)
}

func getRealExposure(id string, windowHours int) (*Exposure, []byte, error) {
	q := url.Values{}
	if windowHours > 0 {
		q.Set("window_hours", strconv.Itoa(windowHours))
	}
	body, err := brokerGetBody("GetExposure", withQuery("/vulnerabilities/"+url.PathEscape(id)+"/exposure", q))
	if err != nil {
		return nil, nil, err
	}
	return decodeInto[Exposure]("GetExposure", body)
}

func getRealSbom(digest, source string, limit int, after int64) (*SbomPage, []byte, error) {
	q := url.Values{}
	if source != "" {
		q.Set("source", source)
	}
	if limit > 0 {
		q.Set("limit", strconv.Itoa(limit))
	}
	if after > 0 {
		q.Set("after", strconv.FormatInt(after, 10))
	}
	body, err := brokerGetBody("GetSbom", withQuery("/images/"+url.PathEscape(digest)+"/sbom", q))
	if err != nil {
		return nil, nil, err
	}
	return decodeInto[SbomPage]("GetSbom", body)
}

func getRealSbomCycloneDX(digest, source string) ([]byte, error) {
	q := url.Values{}
	if source != "" {
		q.Set("source", source)
	}
	resp, err := brokerGet(withQuery("/images/"+url.PathEscape(digest)+"/sbom/cyclonedx", q))
	if err != nil {
		return nil, err
	}
	defer func() {
		if closeErr := resp.Body.Close(); closeErr != nil {
			log.Error().Err(closeErr).Msg("GetSbomCycloneDX: Error closing response body")
		}
	}()
	body, err := io.ReadAll(io.LimitReader(resp.Body, maxCycloneDXBytes+1))
	if err != nil {
		return nil, fmt.Errorf("GetSbomCycloneDX: reading response body: %w", err)
	}
	switch resp.StatusCode {
	case http.StatusOK:
		if len(body) > maxCycloneDXBytes {
			return nil, fmt.Errorf("GetSbomCycloneDX: document larger than %d MiB; page it with 'images sbom' instead", maxCycloneDXBytes>>20)
		}
		return body, nil
	case http.StatusNotFound:
		return nil, &NotFoundError{Message: string(body)}
	default:
		msg := string(body)
		if len(msg) > 200 {
			msg = msg[:200]
		}
		return nil, fmt.Errorf("GetSbomCycloneDX: broker returned HTTP %d: %s", resp.StatusCode, msg)
	}
}
